ironfish-zkp/src/circuits/util.rs

```diff
@@ -0,0 +1,91 @@
+use std::slice;
+
+use bellman::{
+    gadgets::boolean::{self, AllocatedBit, Boolean},
+    ConstraintSystem, SynthesisError,
+};
+use ff::PrimeField;
+use zcash_proofs::circuit::ecc::EdwardsPoint;
+
+pub fn hash_asset_to_preimage<CS: bellman::ConstraintSystem<bls12_381::Scalar>>(
+    cs: &mut CS,
+    name: [u8; 32],
+    chain: [u8; 32],
+    network: [u8; 32],
+    g_d: EdwardsPoint,
+    pk_d: EdwardsPoint,
+    nonce: u8,
+) -> Result<Vec<boolean::Boolean>, SynthesisError> {
+    let mut combined_preimage = vec![];
+
+    combined_preimage.extend(g_d.repr(cs.namespace(|| "booleanize g_d"))?); // 32
+    combined_preimage.extend(pk_d.repr(cs.namespace(|| "booleanize pk_d"))?); // 32
+
+    let name_bits =
+        slice_into_boolean_vec_le(cs.namespace(|| "booleanize name"), Some(&name), 32 * 8)?;
+    assert_eq!(name_bits.len(), 32 * 8);
```

Comment:
Both of these are good suggestions. It's probably better to throw an error here, and we can do that from the `slice_..` fn, although we're going to do a bit more refactoring on that still

Reply:
It might be worth actually discussing if network nodes should crash if particular invariants are violated. If somehow someone is able to violate invariant the risk becomes present to potentially sneak invalid things into proofs which is on the more catastrophic end of things that could go wrong.